### Model application to dataframe
This is a little difficult to follow so here's an explanation:

Step 1: run the model against the entire dataframe to collect the topics
    
Step 2: take this model and apply it back to the dataframe to assign most likely topic to each case (we want the topic # and its dot product)
                                                                                                    
Step 3: make a dictionary of the components that make up each topic from the original model
    
Step 4: use this dictionary to "look up" the topic components and apply those to the dataframe

Step 5: Getting data together for visualization!

In [3]:
import pandas as pd
import re

In [4]:
##########################################  modeling imports  #######################################################
from sklearn.feature_extraction.text import TfidfVectorizer, TfidfTransformer
from sklearn.decomposition import NMF
#from sklearn.preprocessing import Normalizer

In [6]:
df = pd.read_pickle("full_proj_lemmatized3.pickle")

In [7]:
df.head(5)

caseurl  \
15000  http://caselaw.findlaw.com/us-supreme-court/38...   
15001  http://caselaw.findlaw.com/us-supreme-court/38...   
15002  http://caselaw.findlaw.com/us-supreme-court/38...   
15003  http://caselaw.findlaw.com/us-supreme-court/38...   
15004  http://caselaw.findlaw.com/us-supreme-court/38...   

                                               casetitle  years  \
15000  ALUMINUM CO. OF AMERICA v. UNITED STATES,  382...   1965   
15001  JONES & LAUGHLIN STEEL CORP. v. GRIDIRON STEEL...   1965   
15002             JORDAN v. SILVER,  381 U.S. 415 (1965)   1965   
15003           KADANS v. DICKERSON,  382 U.S. 22 (1965)   1965   
15004  METROMEDIA, INC. v. AMERICAN SOCIETY OF COMPOS...   1965   

                                                    case  \
15000    United States Supreme Court JOBE v. CITY OF ...   
15001    United States Supreme Court JONES & LAUGHLIN...   
15002    United States Supreme Court JORDAN v. SILVER...   
15003    United States Supreme Court KADANS v. DICKER...   
15004    United States Supreme Court KASHARIAN v. MET...   

                                                     lem  
15000  jobe appellant appellee want substantial proba...  
15001  gridiron civ proc extend limit expire saturday...  
15002  corker gruskin selvin appellant phill appellee...  
15003  appellant parraguirre appellees want jurisdict...  
15004  appellant conover appellees want jurisdiction ...

In [8]:
df.ix[15000, "caseurl"]

'http://caselaw.findlaw.com/us-supreme-court/382/12.html'

## Step 1: Run model against entire dataframe (as a corpus)
Think of it like this: We need to find the themes across the entire set of documents (over 23,000 in all), so how else would we do this than stacking every document together as a reservoir to extract information out of?

In [231]:
def nmf_mod(corp ):
    df = .80
    n_topics = 30
    n_features = 2000
    n_top_words = 40
    
    # Use tf-idf features for NMF.
    print("Extracting tf-idf features for NMF...")
    tfidf_vectorizer = TfidfVectorizer(max_df=df, min_df=5, # ngram_range=(1,2), #max_features=n_features,
                                       stop_words='english')

    tfidf = tfidf_vectorizer.fit_transform(corp)


    # Fit the NMF model
    print("Fitting the NMF model with tf-idf features, "
          "n_topics= %d, n_topic_words= %d, n_features= %d..."
          % (n_topics, n_top_words, n_features))

    nmf = NMF(n_components=n_topics, random_state=2, alpha=.1, l1_ratio=.5).fit(tfidf)
    
    print("\nTopics in NMF model:")
    tfidf_feature_names = tfidf_vectorizer.get_feature_names()
    #return print_top_words(nmf, tfidf_feature_names, n_top_words) 
    return tfidf,nmf

In [232]:
tfidf, nmf_mod_test = nmf_mod(df.lem)

Extracting tf-idf features for NMF...
Fitting the NMF model with tf-idf features, n_topics= 30, n_topic_words= 40, n_features= 2000...

Topics in NMF model:


## Step 2: Applying the model back to the dataframe 
NMF (as well as other types of topic modeling) returns a matrix of likelihoods that a particular document fits in Topic 1, 2, etc. Unlike LDA, __An NMF matrix does not contain probabilities of inclusion, but rather the dot product of two matrices__. Don't worry about the (linear algebra) details, just imagine that we need to find the biggest number in this matrix and return the index of that. 

In [233]:
out =nmf_mod_test.transform(tfidf)
out[49] #verified that each of these is different 

array([ 0.        ,  0.        ,  0.01743965,  0.        ,  0.        ,
        0.00812628,  0.03701182,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.05023011,  0.00074146,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.00029258,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ])

__Returning these as a Series__
It's easy to run the model against a column of the dataframe, return it as a series, and append that series as a new column. (But remember not to sort if you do this because you need the order to stay the same, obviously).

In [234]:
import operator
topics = []
for item in out:
    max_index, max_value = max(enumerate(item), key=operator.itemgetter(1))
    topics.append(max_index) 
    
df["topicnumber"] = pd.Series(topics, index=df.index)    

In [235]:
topics_likelihood = []
for item in out:
    max_index, max_value = max(enumerate(item), key=operator.itemgetter(1))
    topics_likelihood.append(max_value)
    
df["strengthoftopic"] = pd.Series(topics_likelihood, index=df.index)     

In [236]:
df.topicnumber.value_counts() #let's make sure this is a good model...

16    2053
0     1567
22    1539
27    1457
3     1343
19    1242
10    1039
29     944
20     943
4      860
11     764
21     734
1      680
15     652
18     589
6      559
5      555
26     552
14     547
23     542
13     540
7      458
8      450
9      449
17     434
2      405
24     401
28     377
12     366
25     227
Name: topicnumber, dtype: int64

### Step 3: Creating dictionary of topic components
There's probably an easier way to do this, but I haven't found one. I'm running the model function again (random state will get the same results as before) but this time creating a topic words feature space to "look up" in my dataframe.

In [221]:
def nmf_topics_dict(corp, n_topics):
    df = .80
    n_top_words = 40
    
    tfidf_vectorizer = TfidfVectorizer(max_df=df, min_df=5,# ngram_range=(1,2), #max_features=n_features,
                                       stop_words='english')

    tfidf = tfidf_vectorizer.fit_transform(corp)
    nmf = NMF(n_components=n_topics, random_state=2, alpha=.1, l1_ratio=.5).fit(tfidf)
    tfidf_feature_names = tfidf_vectorizer.get_feature_names()
      
    topic_dict = {}
    for topic_idx, topic in enumerate(nmf.components_):
        topic_dict[topic_idx] = ", ".join([tfidf_feature_names[i] \
                                    for i in topic.argsort()[:-n_top_words - 1:-1]])
    return topic_dict

In [222]:
# I tested a few different topic distributions, 30 was optimal
nmf_words_30 = nmf_topics_dict(df.lem, 30) #dict object

In [223]:
nmf_words_30

{0: 'error, defendant, sup, deliver, judgment, statute, mckenna, bring, record, messrs, prosecute, assignment, recover, favor, render, pass, verdict, proceeding, demurrer, cost, exception, try, sustain, assign, appearance, overrule, allow, purpose, assess, make, come, involve, result, raise, memorandum, contrary, validity, contention, mandamus, approve',
 1: 'vacate, remand, consideration, judgment, solicitor, moot, reconsideration, proceeding, suggestion, respondent, reason, app, ninth, seventh, pauperis, forma, disposition, hearing, inconsistent, result, appellate, record, basis, appropriate, merit, eighth, habeas, consistent, outright, complaint, supra, relation, manoli, dismissal, finan, ordman, gillen, wulf, consider, timely',
 2: 'want, substantial, report, appellate, app, properly, rhyne, consideration, solicitor, grey, quel, appellees, kaufmann, derengoski, probable, bourke, melaniphy, sybert, tighe, torina, londerholm, reversed, jurisdictional, botter, stedman, toch, hickey, u

In [224]:
import json
with open('finaliteration_topics.json', 'w') as fp:
    json.dump(nmf_words_30, fp)

## Step 4: Looking up topic words for each item in dataframe

In [237]:
def word_lookup(num):
    return nmf_words_30.get(num)

In [238]:
df["words"] = df.topicnumber.apply(word_lookup)

In [239]:
df.ix[15017,"words"] # This cell and the one below verifies that it worked

'forma, pauperis, proceed, reason, examination, submit, reverse, docketing, frivolous, ninth, hearing, compliance, supra, issue, clerk, habeas, record, respondent, extraordinary, consideration, relief, filing, vacate, docket, transcript, forth, request, app, set, noncriminal, status, allow, mahorner, unless, argument, faircloth, result, denial, appellate, fender'

In [240]:
df.ix[14972,"lem"]

'appellant licensed physician convict accessory married person information advice prevent conception follow examination prescribe contraceptive device wife statute make person article prevent conception appellant claim accessory statute apply violate fourteenth appellate judgment appellant standing a'

In [14]:
df.ix[15017,"caseurl"]

'http://caselaw.findlaw.com/us-supreme-court/380/145.html'

In [9]:
df.to_pickle("full_project_modelled_final.pickle")

In [10]:
df = pd.read_pickle("full_project_modelled_final.pickle")

## Step 5: Putting data together for visualization
I'm making a brushable area chart with D3. This visualization requires a datapoint for every topic for every year, so we have to do some pivoting to make that happen.

In [119]:
# some topics were extremely similar and at the suggestion of my instructors,
# for the sake of the visualization, I have condensed the topics to 20

def topic_condenser(topicnum):
    if topicnum == 20:
        return 24
    if topicnum == 25:
        return 1
    if topicnum == 2:
        return 12
    if topicnum == 27:
        return 26
    if topicnum == 18 or topicnum == 5:
        return 29
    if topicnum == 8 or topicnum == 22:
        return 7
    if topicnum == 15:
        return 16
    if topicnum == 9:
        return 14
    if topicnum == 19:
        return 3
    else: 
        return topicnum
df["condensedtopics"] = df.topicnumber.apply(topic_condenser)

In [120]:
# doing some research on the not so obvious topics
df = df[df["topicnumber"] != 2]
#df_16.ix[15065, "caseurl"]
df_16

caseurl  \
15006  http://caselaw.findlaw.com/us-supreme-court/38...   
15042  http://caselaw.findlaw.com/us-supreme-court/38...   
15049  http://caselaw.findlaw.com/us-supreme-court/38...   
15053  http://caselaw.findlaw.com/us-supreme-court/38...   
15063  http://caselaw.findlaw.com/us-supreme-court/38...   
15090  http://caselaw.findlaw.com/us-supreme-court/38...   
15127  http://caselaw.findlaw.com/us-supreme-court/38...   
15133  http://caselaw.findlaw.com/us-supreme-court/38...   
15136  http://caselaw.findlaw.com/us-supreme-court/38...   
15137  http://caselaw.findlaw.com/us-supreme-court/38...   
15146  http://caselaw.findlaw.com/us-supreme-court/38...   
15148  http://caselaw.findlaw.com/us-supreme-court/38...   
15172  http://caselaw.findlaw.com/us-supreme-court/38...   
15175  http://caselaw.findlaw.com/us-supreme-court/38...   
15184  http://caselaw.findlaw.com/us-supreme-court/38...   
15202  http://caselaw.findlaw.com/us-supreme-court/38...   
15228  http://caselaw.findlaw.com/us-supreme-court/38...   
15230  http://caselaw.findlaw.com/us-supreme-court/38...   
15251  http://caselaw.findlaw.com/us-supreme-court/38...   
15265  http://caselaw.findlaw.com/us-supreme-court/38...   
15272  http://caselaw.findlaw.com/us-supreme-court/38...   
15277  http://caselaw.findlaw.com/us-supreme-court/38...   
15280  http://caselaw.findlaw.com/us-supreme-court/38...   
15300  http://caselaw.findlaw.com/us-supreme-court/38...   
15386  http://caselaw.findlaw.com/us-supreme-court/38...   
15387  http://caselaw.findlaw.com/us-supreme-court/38...   
15392  http://caselaw.findlaw.com/us-supreme-court/38...   
15397  http://caselaw.findlaw.com/us-supreme-court/38...   
15437  http://caselaw.findlaw.com/us-supreme-court/38...   
15445  http://caselaw.findlaw.com/us-supreme-court/38...   
...                                                  ...   
14507  http://caselaw.findlaw.com/us-supreme-court/37...   
14517  http://caselaw.findlaw.com/us-supreme-court/37...   
14536  http://caselaw.findlaw.com/us-supreme-court/37...   
14553  http://caselaw.findlaw.com/us-supreme-court/37...   
14568  http://caselaw.findlaw.com/us-supreme-court/37...   
14584  http://caselaw.findlaw.com/us-supreme-court/37...   
14617  http://caselaw.findlaw.com/us-supreme-court/37...   
14618  http://caselaw.findlaw.com/us-supreme-court/37...   
14659  http://caselaw.findlaw.com/us-supreme-court/37...   
14666  http://caselaw.findlaw.com/us-supreme-court/37...   
14670  http://caselaw.findlaw.com/us-supreme-court/37...   
14722  http://caselaw.findlaw.com/us-supreme-court/37...   
14756  http://caselaw.findlaw.com/us-supreme-court/37...   
14758  http://caselaw.findlaw.com/us-supreme-court/37...   
14794  http://caselaw.findlaw.com/us-supreme-court/37...   
14815  http://caselaw.findlaw.com/us-supreme-court/37...   
14851  http://caselaw.findlaw.com/us-supreme-court/37...   
14885  http://caselaw.findlaw.com/us-supreme-court/37...   
14898  http://caselaw.findlaw.com/us-supreme-court/38...   
14922  http://caselaw.findlaw.com/us-supreme-court/38...   
14926  http://caselaw.findlaw.com/us-supreme-court/38...   
14930  http://caselaw.findlaw.com/us-supreme-court/38...   
14938  http://caselaw.findlaw.com/us-supreme-court/37...   
14945  http://caselaw.findlaw.com/us-supreme-court/38...   
14955  http://caselaw.findlaw.com/us-supreme-court/38...   
14957  http://caselaw.findlaw.com/us-supreme-court/37...   
14971  http://caselaw.findlaw.com/us-supreme-court/38...   
14982  http://caselaw.findlaw.com/us-supreme-court/38...   
14985  http://caselaw.findlaw.com/us-supreme-court/37...   
14994  http://caselaw.findlaw.com/us-supreme-court/37...   

                                               casetitle  years  \
15006  FLORIDA EAST COAST RAILWAY CO. v. UNITED STATE...   1965   
15042              O'CONNOR v. OHIO,  382 U.S. 19 (1965)   1965   
15049  PRICE v. STATE ROAD COMMISSION OF WEST VIRGINI...   1965   
15053             BURNETTE v. DAVIS,  382 U.S. 42 (1965)   1

In [124]:
df_details = pd.read_csv("detailsford3.csv", encoding = 'iso-8859-1')
df_details.columns = ["condensedtopics", "topicname", "title", "exampleURL", "leadpp", "topicwords"]
df_details

condensedtopics                                         topicname  \
0                 4                                    Interstate Law   
1                16                     Civil Rights - discrimination   
2                24                                         Municipal   
3                 1             Right to Defense Regardless of Income   
4                12                                        Monopolies   
5                23                                       Free Speech   
6                26                              Stocks & Fair Values   
7                29          Torts - a civil wrong that caused a loss   
8                10                                    Workers Unions   
9                 6  Jurisdiction - states, immigration, reservations   
10               11                    Conspiracies & Organized Crime   
11               17                                     Crimes - mail   
12                7                    Violent Crimes & Death Penalty   
13                0                    Criminal Activity - nonviolent   
14               14                                        Bankruptcy   
15               13                           Domestic & Trust Issues   
16               21                 Civil Rights - search and seizure   
17               28          Tenants Rights - business and individual   
18                3                                        Negligence   

                                         title  \
0     KOEHRING CO. v. HYDE CONSTR. CO., (1966)   
1           LOUISIANA v. UNITED STATES, (1965)   
2            CAMARA v. MUNICIPAL COURT, (1967)   
3                 GIDEON v. WAINWRIGHT, (1963)   
4   UNITED STATES v. UNITED SHOE CORP., (1968)   
5             MEMOIRS v. MASSACHUSETTS, (1966)   
6    FRIBOURG NAV. CO. v. COMMISSIONER, (1966)   
7      UNITED STATES v. ATLAS INS. CO., (1965)   
8                 LABOR BOARD v. BROWN, (1965)   
9                       DURFEE v. DUKE, (1963)   
10             FORMAN v. UNITED STATES, (1960)   
11             SINGER v. UNITED STATES, (1965)   
12            BUMPER v. NORTH CAROLINA, (1968)   
13     ORNELAS et al. v. UNITED STATES, (1996)   
14             PHELPS v. UNITED STATES, (1975)   
15                 STANLEY v. ILLINOIS, (1972)   
16                        MAPP v. OHIO, (1961)   
17     MILLINERY CORP. v. COMMISSIONER, (1956)   
18                VAN DUSEN v. BARRACK, (1964)   

                                           exampleURL  \
0   http://caselaw.findlaw.com/us-supreme-court/38...   
1   http://caselaw.findlaw.com/us-supreme-court/38...   
2   http://caselaw.findlaw.com/us-supreme-court/38...   
3   http://caselaw.findlaw.com/us-supreme-court/37...   
4   http://caselaw.findlaw.com/us-supreme-court/39...   
5   http://caselaw.findlaw.com/us-supreme-court/38...   
6   http://caselaw.findlaw.com/us-supreme-court/38...   
7   http://caselaw.findlaw.com/us-supreme-court/38...   
8   http://caselaw.findlaw.com/us-supreme-court/38...   
9   http://caselaw.findlaw.com/us-supreme-court/37...   
10  http://caselaw.findlaw.com/us-supreme-court/36...   
11  http://caselaw.findlaw.com/us-supreme-court/38...   
12  http://caselaw.findlaw.com/us-supreme-court/39...   
13  http://caselaw.findlaw.com/us-supreme-court/51...   
14  http://caselaw.findlaw.com/us-supreme-court/42...   
15  http://caselaw.findlaw.com/us-supreme-court/40...   
16  http://caselaw.findlaw.com/us-supreme-court/36...   
17  http://caselaw.findlaw.com/us-supreme-court/35...   
18  http://caselaw.findlaw.com/us-supreme-court/37...   

                                               leadpp  \
0   On March 10, 1964, the Court of Appeals for th...   
1   Pursuant to 42 U.S.C. 1971 (c) the Attorney Ge...   
2   Appellant was charged with violating the San F...   
3   Charged in a Florida State Court with a noncap...   
4   In 1953 the District Court for the District of...   
5   Appellee, the Attorney General of Massachusett...   
6 

In [130]:
df_with_details = pd.merge(df, df_details, how = "inner", on = "condensedtopics")

In [148]:
#temp_df = df_with_details[['years', 'condensedtopics', "topicname", "title", "exampleURL", "leadpp", "topicwords"]]
#temp_df.to_csv("temp.csv")
temp_df = df_with_details[['years', 'condensedtopics']]
temp_df.condensedtopics.value_counts()

16    2705
3     2585
7     2447
29    2088
26    2009
0     1567
24    1344
10    1039
14     996
1      907
4      860
11     764
21     734
6      559
23     542
13     540
17     434
28     377
12     366
Name: condensedtopics, dtype: int64

In [ ]:
#dummy value for each existing topic. Pay no attention to this error.
temp_df["count"] = 1
temp_df

In [150]:
#this condenses each point for the same year into n number of points 
temp_df = temp_df.groupby(["years", "condensedtopics"]).count().reset_index()
temp_df

years  condensedtopics  count
0      1792               21      1
1      1793               24      2
2      1795                7      1
3      1795               10      1
4      1796                3      1
5      1796                4      1
6      1796               21      1
7      1796               29      1
8      1798                4      1
9      1798               21      1
10     1798               24      1
11     1799                0      1
12     1799                3      1
13     1800                4      1
14     1800               16      1
15     1801                4      1
16     1802                3      1
17     1803                0      2
18     1803                7      1
19     1804               24      1
20     1805                6      1
21     1805               11      1
22     1805               14      1
23     1806                3      2
24     1806                6      1
25     1806                7      1
26     1807                7      1
27     1807               11      1
28     1808                6      1
29     1808                7      1
...     ...              ...    ...
2593   2015                1      5
2594   2015                3      6
2595   2015                4      1
2596   2015                7     18
2597   2015               10      3
2598   2015               11      3
2599   2015               12      2
2600   2015               13      3
2601   2015               14      5
2602   2015               16     18
2603   2015               21      7
2604   2015               24      3
2605   2015               26      3
2606   2015               29      7
2607   2016                1      4
2608   2016                3      2
2609   2016                6      6
2610   2016                7     17
2611   2016               10      6
2612   2016               11      4
2613   2016               12      1
2614   2016               13      1
2615   2016               14      3
2616   2016               16      8
2617   2016               17      5
2618   2016               21      2
2619   2016               23      1
2620   2016               24      2
2621   2016               26      3
2622   2016               29      8

[2623 rows x 3 columns]

#### A few (really cool) things are happening in the cell below 

First, we are pivoting to add dummy values for nonexistent 
year/topic points (for ex, there's only 1 case in 1792 but 30 topics, we need 29 points of 0). The topic numbers become column headers first, followed by filling the NaNs with 0's, then we stack the df back to the way it was and reset the index.



In [ ]:
data_fillna = temp_df.pivot_table("count", "years", "condensedtopics").fillna(0).unstack().reset_index()

In [152]:
#we lose the count label column in the previous steps, so we're just renaming it here, and reordering columns based on 
#how they are arranged in the viz csv
data_fillna.columns = ["condensedtopics", "years", "count"]
data_fillna = data_fillna[["years", "condensedtopics", "count"]]

In [153]:
#sort by year
final_data = pd.merge(data_fillna, df_details, how = "inner", on = "condensedtopics")
final_data

years  condensedtopics  count                                 topicname  \
0      1792                0      0            Criminal Activity - nonviolent   
1      1793                0      0            Criminal Activity - nonviolent   
2      1795                0      0            Criminal Activity - nonviolent   
3      1796                0      0            Criminal Activity - nonviolent   
4      1798                0      0            Criminal Activity - nonviolent   
5      1799                0      1            Criminal Activity - nonviolent   
6      1800                0      0            Criminal Activity - nonviolent   
7      1801                0      0            Criminal Activity - nonviolent   
8      1802                0      0            Criminal Activity - nonviolent   
9      1803                0      2            Criminal Activity - nonviolent   
10     1804                0      0            Criminal Activity - nonviolent   
11     1805                0      0            Criminal Activity - nonviolent   
12     1806                0      0            Criminal Activity - nonviolent   
13     1807                0      0            Criminal Activity - nonviolent   
14     1808                0      0            Criminal Activity - nonviolent   
15     1809                0      0            Criminal Activity - nonviolent   
16     1810                0      0            Criminal Activity - nonviolent   
17     1812                0      0            Criminal Activity - nonviolent   
18     1813                0      0            Criminal Activity - nonviolent   
19     1814                0      0            Criminal Activity - nonviolent   
20     1815                0      0            Criminal Activity - nonviolent   
21     1816                0      1            Criminal Activity - nonviolent   
22     1817                0      0            Criminal Activity - nonviolent   
23     1818                0      2            Criminal Activity - nonviolent   
24     1819                0      2            Criminal Activity - nonviolent   
25     1820                0      0            Criminal Activity - nonviolent   
26     1821                0      2            Criminal Activity - nonviolent   
27     1822                0      0            Criminal Activity - nonviolent   
28     1823                0      0            Criminal Activity - nonviolent   
29     1824                0      0            Criminal Activity - nonviolent   
...     ...              ...    ...                                       ...   
4188   1987               28      3  Tenants Rights - business and individual   
4189   1988               28      2  Tenants Rights - business and individual   
4190   1989               28      3  Tenants Rights - business and individual   
4191   1990               28      0  Tenants Rights - business and individual   
4192   1991               28      0  Tenants Rights - business and individual   
4193   1992               28      1  Tenants Rights - business and individual   
4194   1993               28      2  Tenants Rights - business and individual   
4195   1994               28      0  Tenants Rights - business and individual   
4196   1995               28      0  Tenants Rights - business and individual   
4197   1996               28      1  Tenants Rights - business and individual   
4198   1997               28      1  Tenants Rights - business and individual   
4199   1998               28      1  Tenants Rights - business and individual   
4200   1999               28      2  Tenants Rights - business and individual   
4201   2000               28      2  Tenants Rights - business and individual   
4202   2001               28      1  Tenants Rights - business and individual   
4203   2002               28      2  Tenants Rights - business and individual   
4204   2003               28      1  Tenants Rights - business and individual   
4205   2004               28      2  

In [154]:
final_data.sort_values("years", inplace = True, ascending = True)
final_data

years  condensedtopics  count  \
0      1792                0      0   
1554   1792               21      1   
2664   1792                7      0   
444    1792               17      0   
1998   1792               29      0   
2220   1792               12      0   
1332   1792               10      0   
2886   1792               23      0   
2442   1792                6      0   
3996   1792               28      0   
1776   1792               16      0   
888    1792               13      0   
1110   1792               14      0   
3108   1792                4      0   
666    1792                3      0   
3552   1792                1      0   
222    1792               11      0   
3330   1792               26      0   
3774   1792               24      0   
1777   1793               16      0   
1111   1793               14      0   
2443   1793                6      0   
3109   1793                4      0   
445    1793               17      0   
2665   1793                7      0   
223    1793               11      0   
3775   1793               24      2   
3997   1793               28      0   
3331   1793               26      0   
1      1793                0      0   
...     ...              ...    ...   
2884   2015                7     18   
1774   2015               21      7   
2440   2015               12      2   
3550   2015               26      3   
442    2015               11      3   
3772   2015                1      5   
2218   2015               29      7   
4216   2015               28      0   
3994   2015               24      3   
664    2015               17      0   
1330   2015               14      5   
1775   2016               21      2   
2219   2016               29      8   
2441   2016               12      1   
1553   2016               10      6   
2663   2016                6      6   
3995   2016               24      2   
2885   2016                7     17   
1109   2016               13      1   
3107   2016               23      1   
1997   2016               16      8   
887    2016                3      2   
3329   2016                4      0   
665    2016               17      5   
3551   2016               26      3   
443    2016               11      4   
3773   2016                1      4   
221    2016                0      0   
1331   2016               14      3   
4217   2016               28      0   

                                             topicname  \
0                       Criminal Activity - nonviolent   
1554                 Civil Rights - search and seizure   
2664                    Violent Crimes & Death Penalty   
444                                      Crimes - mail   
1998          Torts - a civil wrong that caused a loss   
2220                                        Monopolies   
1332                                    Workers Unions   
2886                                       Free Speech   
2442  Jurisdiction - states, immigration, reservations   
3996          Tenants Rights - business and individual   
1776                     Civil Rights - discrimination   
888                            Domestic & Trust Issues   
1110                                        Bankruptcy   
3108                                    Interstate Law   
666                                         Negligence   
3552             Right to Defense Regardless of Income   
222                     Conspiracies & Organized Crime   
3330                              Stocks & Fair Values   
3774                                         Municipal   
1777                     Civil Rights - discrimination   
1111                                        Bankruptcy   
2443  Jurisdiction - states, immigration, reservations   
3109                                    Interstate Law   
445                                      Crimes - mail   
2665                    Violent Crimes & Death Penalty   
223                     Conspiracies & Organized Crime   
3775                 

In [156]:
#backup file
final_data.to_csv("topicsbyyear.csv", index = False)
final_data.to_csv("year_topic_data2.csv", index = False)

In [ ]:
'''the best part of this viz is the brushing side to side effect. For that, we need total cases for every year
and need no other information'''

data_fillna.groupby("years")["count"].sum().reset_index().to_csv("year_data.csv", index = False)

## _c'est fin!_